buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java

```diff
@@ -0,0 +1,124 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Perform the transformations against the set of RestTests from a given file.
+ */
+public class RestTestTransformer {
+    /**
+     * Transforms a REST test based on the requested {@link RestTestTransform}'s
+     *
+     * @param tests           The REST tests from the same file. Uses linked list so we can easily add to the beginning of the list.
+     * @param transformations The set of transformations to perform against the test
+     * @return the transformed tests
+     */
+    public List<ObjectNode> transformRestTests(LinkedList<ObjectNode> tests, List<RestTestTransform<?>> transformations) {
+        ObjectNode setupSection = null;
+        ObjectNode teardownSection = null;
+
+        // Collect any global setup transformations
+        List<RestTestTransformGlobalSetup> setupTransforms = transformations.stream()
+            .filter(transform -> transform instanceof RestTestTransformGlobalSetup)
+            .map(transform -> (RestTestTransformGlobalSetup) transform)
+            .collect(Collectors.toList());
+
+        // Collect any global setup transformations
+        List<RestTestTransformGlobalTeardown> teardownTransforms = transformations.stream()
+            .filter(transform -> transform instanceof RestTestTransformGlobalTeardown)
+            .map(transform -> (RestTestTransformGlobalTeardown) transform)
+            .collect(Collectors.toList());
+
+        // transform the tests and include the global setup and teardown as part of the transform
+        for (ObjectNode test : tests) {
+            Iterator<Map.Entry<String, JsonNode>> testsIterator = test.fields();
+            while (testsIterator.hasNext()) {
+                Map.Entry<String, JsonNode> testObject = testsIterator.next();
+                String testName = testObject.getKey();
+                if ("setup".equals(testName)) {
+                    setupSection = test;
+                }
+                if ("teardown".equals(testName)) {
+                    teardownSection = test;
+                }
+                Map<String, RestTestTransformByObjectKey> objectKeyFinders = transformations.stream()
+                    .filter(transform -> transform instanceof RestTestTransformByObjectKey)
+                    .map(transform -> (RestTestTransformByObjectKey) transform)
+                    .collect(Collectors.toMap(RestTestTransformByObjectKey::getKeyToFind, transform -> transform));
+
+                transformByObjectKeyName(test, objectKeyFinders);
+            }
+        }
+
+        // transform the global setup
+        if (setupTransforms.isEmpty() == false) {
+            int i = 0;
+            for (RestTestTransformGlobalSetup setupTransform : setupTransforms) {
+                ObjectNode result = setupTransform.transformSetup(setupSection);
+                if (result != null && setupSection == null && i++ == 0) {
```

Comment:
can you put a comment about this expression?
Is the intention that if there is no setupSection, but there is a transformation for a setup, then add a setup?
but it also looks like only one can be added? What if I have two setup transformations ?

Reply:
> can you put a comment about this expression?
will do. 

> Is the intention that if there is no setupSection, but there is a transformation for a setup, then add a setup?
yes, that is the intention (create one if missing)

> but it also looks like only one can be added? What if I have two setup transformations ?
good catch . There is a bug as-is with multiple setup/teardown transforms which will get fixed ... however since there is only at most 1 with this PR it is not tested yet , but will likely run into that sooner then later once i introduce more transforms.